src/NServiceBus.Core.Tests/Msmq/MsmqFailureInfoStorageTests.cs

```diff
@@ -0,0 +1,126 @@
+
+            storage.RecordFailureInfoForMessage(messageId, exception);
+
+            MsmqFailureInfoStorage.ProcessingFailureInfo failureInfo;
+            storage.TryGetFailureInfoForMessage(messageId, out failureInfo);
+
+            Assert.NotNull(failureInfo);
+            Assert.AreEqual(1, failureInfo.NumberOfProcessingAttempts);
+            Assert.AreSame(exception, failureInfo.Exception);
+        }
+
+        [Test]
+        public void When_recording_failure_many_times_should_store_number_of_attempts_and_last_exception()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+            var secondException = new Exception();
+
+            var storage = GetFailureInfoStorage();
+
+            storage.RecordFailureInfoForMessage(messageId, new Exception());
+            storage.RecordFailureInfoForMessage(messageId, secondException);
+
+            MsmqFailureInfoStorage.ProcessingFailureInfo failureInfo;
+            storage.TryGetFailureInfoForMessage(messageId, out failureInfo);
+
+            Assert.NotNull(failureInfo);
+            Assert.AreEqual(2, failureInfo.NumberOfProcessingAttempts);
+            Assert.AreSame(secondException, failureInfo.Exception);
+        }
+
+        [Test]
+        public void When_clearing_failure_should_return_null_on_subsequent_retrieval()
+        {
+            var messageId = Guid.NewGuid().ToString("D");
+
+            var storage = GetFailureInfoStorage();
+
+            storage.RecordFailureInfoForMessage(messageId, new Exception());
+
+            MsmqFailureInfoStorage.ProcessingFailureInfo failureInfo;
+
+            storage.TryGetFailureInfoForMessage(messageId, out failureInfo);
+            Assert.NotNull(failureInfo);
+            
+            storage.ClearFailureInfoForMessage(messageId);
+
+            storage.TryGetFailureInfoForMessage(messageId, out failureInfo);
+            Assert.IsNull(failureInfo);
+        }
+
+        [Test]
+        public void When_recording_more_than_max_number_of_failures_should_remove_least_recently_used_entry()
+        {
+            const int MaxElements = 50;
+            var storage = new MsmqFailureInfoStorage(maxElements: MaxElements);
+
+            var lruMessageId = Guid.NewGuid().ToString("D");
+
+            storage.RecordFailureInfoForMessage(lruMessageId, new Exception());
+
+            for (var i = 0; i < MaxElements; ++i)
+            {
+                var messageId = Guid.NewGuid().ToString("D");
+                var exception = new Exception();
+
+                storage.RecordFailureInfoForMessage(messageId, exception);
+            }
+
+            MsmqFailureInfoStorage.ProcessingFailureInfo failureInfo;
+            storage.TryGetFailureInfoForMessage(lruMessageId, out failureInfo);
+
+            Assert.IsNull(failureInfo);
+        }
+
+        [Test]
+        public void When_recording_more_than_max_number_of_failures_many_times_should_remove_least_recently_used_entry()
```

Comment:
I'm not really getting this test. What's the essential difference between the one above? The test states `should_remove_least_recently_used_entry` but I don't see any assertion for that?